plan/aggregation_eliminate.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+// // Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+)
+
+// aggEliminater tries to elimiate max/min aggregate function.
+// For SQL like `select max(c) from t;`, we could optimize it to `select c from t order by c desc limit 1;`.
+// For SQL like `select min(c) from t;`, we could optimize it to `select c from t order by c limit 1;`.
+type aggEliminater struct {
+	ctx context.Context
+}
+
+func (a *aggEliminater) optimize(p LogicalPlan, ctx context.Context) (LogicalPlan, error) {
+	a.ctx = ctx
+	return a.eliminateAgg(p), nil
+}
+
+// Try to convert max/min to Limit+Sort operators.
+func (a *aggEliminater) eliminateAgg(p LogicalPlan) LogicalPlan {
+	if agg, ok := p.(*LogicalAggregation); ok {
+		// We only consider case with single max/min function.
+		if len(agg.AggFuncs) != 1 {
+			return p
+		}
+		f := agg.AggFuncs[0]
+		if f.GetName() != ast.AggFuncMax && f.GetName() != ast.AggFuncMin {
+			return p
+		}
+
+		// Add Sort and Limit operators.
+		// For max function, the sort order should be desc.
+		desc := f.GetName() == ast.AggFuncMax
+		// Compose Sort operator.
+		sort := Sort{}.init(a.ctx)
+		sort.ByItems = append(sort.ByItems, &ByItems{f.GetArgs()[0], desc})
+		sort.SetSchema(p.Schema().Clone())
```

Comment:
the schema of sort should be set as its child's

Reply:
So as the limit?

Reply:
@hanfei1991